librato/librato.go

```diff
@@ -0,0 +1,148 @@
+package librato
+
+import (
+	"bytes"
+	"encoding/json"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/nyaruka/courier/utils"
+	"github.com/sirupsen/logrus"
+)
+
+// Default is our default librato collector
+var Default *Sender
+
+// NewSender creates a new librato Sender with the passed in parameters
+func NewSender(waitGroup *sync.WaitGroup, username string, token string, source string, timeout time.Duration) *Sender {
+	return &Sender{
+		waitGroup: waitGroup,
+		stop:      make(chan bool),
+
+		buffer:   make(chan gauge, 1000),
+		username: username,
+		token:    token,
+		source:   source,
+		timeout:  timeout,
+	}
+}
+
+// AddGauge can be used to add a new gauge to be sent to librato
+func (c *Sender) AddGauge(name string, value float64) {
+	// if no librato configured, return
+	if c == nil {
+		return
+	}
+
+	// our buffer is full, log an error but continue
+	if len(c.buffer) >= cap(c.buffer) {
```

Comment:
Maybe, but the idea is to avoid mutexes as much as possible with channels. We really should never reach this point as the buffer size should keep us ahead. If for some reason we do go beyond this, what will happen is the push unto the channel will block until our other goroutine flushes, so it isn't the end of the world.